### PR TITLE
PLANNER-1978 Score-related deprecations

### PR DESCRIPTION
**Pull request by triceo**
Some of these changes are just a little bit invasive. Please see code review comments.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java [hunk 1]
```diff
@@ -60,15 +63,22 @@
...
      */
-    Score_ toInitializedScore();
+    @Deprecated
+    default Score_ toInitializedScore() {
+        if (isSolutionInitialized()) {
+            return (Score_) this;
+        } else {
+            return withInitScore(0);
+        }
+    }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:75]: I was able to replace methods on the individual `Score` implementations with this one.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java [hunk 2]
```diff
@@ -155,6 +165,22 @@
...
+     *
+     * @return true if the hard score is 0 or higher and the {@link #getInitScore()} is 0.
+     */
+    default boolean isFeasible() {
+        /*
+         * This exception will only be thrown for custom scores that did not implement FeasibilityScore.
+         * TODO Safe to remove this default implementation once FeasibilityScore has been removed in 8.0.
+         */
+        throw new UnsupportedOperationException();
+    }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:182]: Noteworthy, I anticipate comments here.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java [hunk 3]
```diff
@@ -60,15 +63,22 @@
...
+            return withInitScore(0);
+        }
+    }
 
     /**
      * For example {@code 0hard/-8soft} with {@code -7} returns {@code -7init/0hard/-8soft}.
      *
      * @param newInitScore always negative (except in statistical calculations), 0 if all planning variables are initialized
      * @return equals score except that {@link #getInitScore()} is set to {@code newInitScore}
-     * @throws IllegalStateException if the original {@link #getInitScore()} is not 0
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:71]: Technically this is a backwards incompatible change. Used to throw, doesn't anymore. I think it should be fine.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:71]: EasyScoreDirector and IncrementalScoreDirector should fail fast if the EasyScoreCalculator etc return a Score with a non zero initScore. Before they did, now they don't I believe.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:71]: I'd add a fail-fast check in both those directors, if the score returned by calculateScore() has a non-zero initScore.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java [hunk 4]
```diff
@@ -155,6 +165,22 @@
      */
     boolean isCompatibleArithmeticArgument(Score otherScore);
 
+    /**
+     * A {@link PlanningSolution} is feasible if it has no broken hard constraints
+     * and {@link #isSolutionInitialized()} is true.
+     *
+     * Simple scores ({@link SimpleScore}, {@link SimpleLongScore}, {@link SimpleBigDecimalScore}) are always feasible.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:172]: not sure if it's worth adding, but "(unless their {@link #getInitScore()} isn't 0)"

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java (hunk 2 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/Score.java:182]: I am fine with this approach. When I added toShortString() in FEB 2017 I should have probably done it this way too with a default method until 8.0 is.

Please add this in the list of PLANNER-750 TODO's, to remove make this method abstract again.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/comparator/NaturalScoreComparator.java [hunk 5]
```diff
@@ -23,7 +23,10 @@
 
 /**
  * Compares by the natural order of {@link Score}s.
+ * 
+ * @deprecated If you need this, it is trivial to implement.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/comparator/NaturalScoreComparator.java:27]: "it is trivial to implement" says the rocket engineer on how to build a V2 missle :)
Mention mention how to build a comparator on natural comparison? Especially in the upgrade recipe, to save users from have to look that up on SO.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/comparator/NaturalScoreComparator.java:27]: Ah, its just Comparable::compareTo apparently?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/buildin/hardmediumsoft/HardMediumSoftScoreDefinition.java [hunk 6]
```diff
@@ -21,11 +21,13 @@
...
 import org.optaplanner.core.config.score.trend.InitializingScoreTrendLevel;
-import org.optaplanner.core.impl.score.definition.AbstractFeasibilityScoreDefinition;
+import org.optaplanner.core.impl.score.definition.AbstractScoreDefinition;
+import org.optaplanner.core.impl.score.definition.ScoreDefinition;
 import org.optaplanner.core.impl.score.inliner.ScoreInliner;
 import org.optaplanner.core.impl.score.trend.InitializingScoreTrend;
 
-public class HardMediumSoftScoreDefinition extends AbstractFeasibilityScoreDefinition<HardMediumSoftScore> {
+public class HardMediumSoftScoreDefinition extends AbstractScoreDefinition<HardMediumSoftScore>
+        implements ScoreDefinition<HardMediumSoftScore> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/buildin/hardmediumsoft/HardMediumSoftScoreDefinition.java:30]: AbstractScoreDefinition implements ScoreDefinition, no need to mention it twice. Remove the "implements".
Same in other files.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/definition/AbstractBendableScoreDefinition.java [hunk 7]
```diff
@@ -18,8 +18,8 @@
 
 import org.optaplanner.core.api.score.FeasibilityScore;
 
-public abstract class AbstractBendableScoreDefinition<S extends FeasibilityScore<S>>
-        extends AbstractFeasibilityScoreDefinition<S> {
+public abstract class AbstractBendableScoreDefinition<S extends FeasibilityScore<S>> extends AbstractScoreDefinition<S>
+        implements ScoreDefinition<S> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/definition/AbstractBendableScoreDefinition.java:22]: same here, AbstractScoreDefinition already implements it

**Pull request merged**